AssetRipper.Library/Exporters/Audio/AudioMixerExportCollection.cs

```diff
@@ -0,0 +1,272 @@
+			
+			// append an Attenuation effect to a group if it has not yet got one,
+			// as Unity doesn't store Attenuation effect if it is the last.
+			foreach (var group in groups)
+			{
+				if (!groupsWithAttenuation.Contains(group))
+				{
+					var effect = virtualFile.CreateAsset<IAudioMixerEffectController>(ClassIDType.AudioMixerEffectController);
+					effect.ObjectHideFlags = HideFlags.HideInHierarchy | HideFlags.HideInInspector;
+					var effectPPtr = new PPtr_AudioMixerEffectController_();
+					effectPPtr.CopyValues(effect.SerializedFile.CreatePPtr(effect));
+					group.Effects_C243.Add(effectPPtr);
+					AddAsset(effect);
+					
+					effect.EffectID_C244.CopyValues((GUID)UnityGUID.NewGuid());
+					effect.EffectName_C244.String = "Attenuation";
+				}
+			}
+
+			// collect snapshots
+			
+			for (int i = 0; i < mixer.Snapshots_C241.Count; i++)
+			{
+				var snapshotPPtr = mixer.Snapshots_C241[i];
+				var snapshot = snapshotPPtr.GetAsset(mixer.SerializedFile) as IAudioMixerSnapshotController;
+				AddAsset(snapshot);
+				
+				var snapshotConstant = constants.Snapshots[i];
+				for (int j = 0; j < snapshotConstant.Values.Length; j++)
+				{
+					if (indexToGUID.TryGetValue((uint)j, out var valueGUID))
+					{
+						snapshot.FloatValues_C245[valueGUID] = snapshotConstant.Values[j];
+					}
+					else
+					{
+						Logger.Warning(LogCategory.Export, $"Snapshot({snapshot.Name_C245}) value #{j} has no binding parameter");
+					}
+				}
+
+				for (int j = 0; j < snapshotConstant.TransitionIndices.Length; j++)
+				{
+					uint paramIndex = snapshotConstant.TransitionIndices[j];
+					int transitionType = (int)snapshotConstant.TransitionTypes[j];
+					if (indexToGUID.TryGetValue(paramIndex, out var paramGUID))
+					{
+						snapshot.TransitionOverrides_C245[paramGUID] = transitionType;
+					}
+					else
+					{
+						Logger.Warning(LogCategory.Export, $"Snapshot({snapshot.Name_C245}) transition #{paramIndex} has no binding parameter");
+					}
+				}
+			}
+			
+			// generate exposed parameters
+
+			for (int i = 0; i < constants.ExposedParameterIndices.Length; i++)
+			{
+				uint paramIndex = constants.ExposedParameterIndices[i];
+				uint paramNameCrc = constants.ExposedParameterNames[i];
+				if (indexToGUID.TryGetValue(paramIndex, out var paramGUID))
+				{
+					var exposedParam = mixer.ExposedParameters_C241.AddNew();
+					exposedParam.Guid.CopyValues(paramGUID);
+					exposedParam.NameString = $"{CrcUtils.ReverseDigestAscii(paramNameCrc)}";
+				}
+				else
+				{
+					Logger.Warning(LogCategory.Export, $"Exposed parameter #{paramIndex} has no binding parameter");
+				}
+			}
+			
+			// complete mixer controller
+			
+			var groupView = mixer.AudioMixerGroupViews_C241.AddNew();
+			groupView.NameString = "View";
+			foreach (var group in groups)
+			{
+				groupView.Guids.Add(group.GroupID_C243);
+			}
+			mixer.CurrentViewIndex_C241 = 0;
+			mixer.TargetSnapshot_C241.CopyValues(mixer.StartSnapshot_C241);
+		}
```

Comment:
You do a lot of stuff in this method. I would prefer it be broken down into more manageable chunks. Also, I'm concerned about the modifications you're doing here and how that would affect things if someone ripped the assets twice. In the future, I want to have an intermediate step after loading for asset processing.

Reply:
I've pushed cleaner code. It's relatively simpler to split into load&process phases.

Reply:
> In the future, I want to have an intermediate step after loading for asset processing.

When I said this, I wasn't actually telling you to do anything. I was referring to the changes I've been doing in the `AssetRipper.Assets` project.